dilated-file-transfer.md

```diff
@@ -0,0 +1,513 @@
+
+* Overall versioning considerations
+
+Versioning is hard.
+
+The existing file-transfer protocol does not include versioning.
+Luckily, the mailbox protocol _itself_ allows for "application versioning" messages; this gives us an "out" here, effectively letting us send a "T minus 1" message via the "application version" dict.
+
+We do not have another such escape hatch (i.e. "T minus 2"), so if we get it wrong (again) then we potentially have a painful upgrade path.
+
+Currently, a file-transfer peer that sends zero version data is assumed to be "classic".
+A file-transfer peer supporting Dilation and this new protocol sends `"transfer": {...}` as per the  "Version Negotiation" section above.
+We include a `"version"` key in that dict so that this transfer protocol may be extended in the future (see "Protocol Expansion Exercises" below).
+
+Additionally, a `"features"` key is included in that information.
+Although related, this is somewhat orthogonal to "versions".
+That is, a peer may _know how to parse_ some (newer) version of this protocol but may still wish to _not_ support (or use) a particular feature.
+
+
+* message encoding
+
+While `msgpack` is mentioned above, there are several other binary-supporting libraries worth considering.
+These are (in no particular order) at least: CBOR or flatbuffers or protocolbuffers or cap'n'proto
+
+We could also still decide to simply use JSON, as that is also unambiguous.
+Although it enjoys widespread support, JSON suffers from a lack of 64-bit integers (because JavaScript only supports 2^53 integers) and doesn't support a byte-string type (forcing e.g. any binary data to be encoded in base64 or similar).
+
+preliminary conclusion: msgpack.
+
+
+* file naming
+
+Sending a single file like `/home/meejah/Documents/Letter.docx` gets a filename `Letter.docx`
+Sending a whole directory like `/home/meejah/Documents/` would result in a directory-offer with basedir `Documents` and some number of files (possibly with sub-paths).
+
+This does NOT offer a client the chance to select "this" and "that" from a Directory offer (however, see the "Protocol Expansion Exercises" section).
+
+Preliminary conclusion: centering around "the thing a human would select" (i.e. "a file" or "a directory") makes the most sense.
+
+
+* streaming data
+
+There is no "finished" message. Maybe there should be? (e.g. the receiving side sends back a hash of the file to confirm it received it properly?)
+
+Does "re-using" the `FileOffer` as a kind of "header" when streaming `DirectoryOffer` contents make sense?
+We need _something_ to indicate the next file etc.
+Preliminary conclusion: it's fine and gives consistent metadata
+
+Do the limits on message size make sense? Should "65KiB" be much smaller, potentially?
+(Given that network conditions etc vary a lot, I think it makes sense for the _spec_ to be somewhat flexible here and "65k" doesn't seem very onerous for most modern devices / computers)
```

Comment:
Different clients/networks might require a different setting. Maybe some optional negotiation for this setting would be needed to allow that.

Reply:
Bear in mind that this will be fragmented differently at the network layer; this size is moreso "how much data does the Dilation layer need to buffer before it can decrypt a record and give it to the higher layer".

The default WebSocket message limit is also around this as well, IIRC. (but same thing applies; it will send frames that are smaller and only deliver "a message" to the application layer).

Reply:
"In general" I would prefer to have as little negotiation of things as possible .. however, clients sending "preferred message size" (or "max message size") is I think what you mean?

For _this_ specification, the considerations will I think largely be around memory usage on the clients (as they have to buffer outgoing messages until an ACK is received in the Dilation layer). Since they also have to hold an entire record in memory at _least_ twice (to decrypt it) but probably more (e.g. compression) clients with small total memory may prefer smaller messages.

Part of the reason for the XXX comments and waffling here is that I'd _rather_ just say in the specification what the message size is (well, "max" because obviously at least the last record for a file will be smaller). However, I also don't want to choose one that's way too small. 65KiB seems "definitely bigger than anyone should want" and also takes guidance from the WebSocket specification (which will in any case limit us in some cases when that is the transport).

Reply:
We should distinguish between maximum message size and file chunk size. From the transit spec (we should apply the same to Dilation, as it entails a transit-like implementation):

> Transit implementations must implement a mechanism to set an upper bound to the message size, which protocols using transit may use. That value should default to 64MiB, which is a common default for WebSockets or HTTP communication.

I don't think we really need negotiation here, we can just hard-code this on a per application-level protocol basis. I think 64kiB as a maximum allowed file chunk size is okay, although implementations are free to chose lower (e.g. to align with the hard disk block size).

Reply:
I don't know what you mean by "file chunk size" but all I've talked about here is maximum _message_ size (that _this_ application protocol should send).

Note that WebSocket _messages_ are different from WebSocket _frames_ (one message may be many frames). WebSocket chose 64KiB as the maximum _frame_ size, but does not specify a maximum  _message_ size (leaving that up to implementations).

Autobahn chose 1MiB as the maximum _message_ size when speaking WAMP. There is no maximum message size in WebSockets unless set by client code (although many example configs use 16MiB). Autobahn's "raw socket" mode imposes 16MiB max message size.

(It appears the text you're quoting above was added by you without a corresponding pull-request or discussion and I don't believe the transit server implements any maximum). Where does HTTP impose any 64MiB limit?

Reply:
> (It appears the text you're quoting above was added by you without a corresponding pull-request or discussion and I don't believe the transit server implements any maximum). Where does HTTP impose any 64MiB limit?

That text was added by me in #16, IIRC. Since the limit is application-protocol defined and may be larger, it is not up to the relay server to enforce it. HTTP of course does not impose any such limit, but most any implementation has some kind of configurable limiting mechanism.

If you are talking about overall message size limits, I don't know why it should be so restrictive in the low kilobytes range. Offering a large file may easily go over it, and I wouldn't see much point in adding yet another chunking layer to the stack.

Reply:
Weird, didn't see #16 in the history ..  is there no merge commit?

"offering a large file may easily go over" -> what do you mean? A directory may have more than 65KiB of names?

I'm fine with just removing this and leaving it up to Dilation and/or client library discretion.

Reply:
Okay .. the spec as written is about "data messages" and how to split up the file data so that wouldn't affect the offers, but maybe that's not clear.

I guess I can just remove anything about maximums -- but it feels like we should say _something_ there, probably, so implementations don't just put the entire content in one message..?

Reply:
Then, just recommend sending the bytes in chunks with a size of 2-64 kiB per message. Make it specific to file bytes, and ignore other message kinds (their limits should be dictated by the Dilation/Transit layer, we'll have to figure that out later). Maybe:

> Messages of kind `5` ("file data bytes") consist solely of file data. The bytes should be sent in chunks, which should not exceed 64 kiB in size per message.

Reply:
Yes, that's what I already have done.

Reply:
No, the word "chunk" is less important to me than the perspective. Since this only applies only to kind 5 messages messages, we care less about the overall message size. I think adding one byte to the max size just to include the header byte in the counting is a bit weird. We could spare this and only talk about the actual payload. I just used "chunks" in an attempt to make that distinction more clear.

Reply:
It must not have been clear if you thought it was talking about the other message sizes..?

Reply:
Okay, I'm resolving this. If there's better choices for fragmenting please re-open / "un-resolve?" but otherwise we'll leave it at max 2**16 for now.